feat(client): JsonSerializable and JsonDeserialized

jason-ha opened the pull request:
Add pair of type filters for JSON based serialization.

`JsonSerializable<T>` produces type representing limitations of serializing `T`. Incompatible elements are transformed to `never` or `SerializationError*` types that original `T` is not assignable to.

`JsonDeserialized<T>` produces type representing result of `T` being serialized and then deserialized.

`JsonSerializable` should eventually replace `@fluidframework/datastore-definitions`'s `Jsonable`. That cannot done be currently as it would be a compile-time breaking change.

[AB#6887](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/6887)

### Additional changes beyond Presence implementation

- New support added for `unknown` as and exactly allowed option.
  - `AllowExactly` option changed to tuple in support.
  - Helper added to detect template literals
  - Helper added to detect index signatures
- Support for branded primitives (`boolean`, `number`, and `string`)
- Additional comments and corrections
- Test coverage for arrays nested in objects

### Supporting Changes

Add standard test infrastructure

jason-ha committed:
feat(client): Json serializable utility types

jason-ha committed:
test(client): Serializable type tests

jason-ha committed:
test(client): class instance test for Serialization utility types

jason-ha committed:
test(client): tests for getter/setter properties and Serialization

jason-ha committed:
test(client): tests for recursive objects

jason-ha committed:
fix(client): Deserialized result for `object`
+ test objects with certain properties

Note some new Deserialized tests are failing

jason-ha committed:
fix(client): Deserialized objects with unsupported values
+ test Replacement cases

jason-ha committed:
test(client): cleanup Deserialize test requiring deserialized type for input

jason-ha committed:
fix(client): Serializable recursion
+ cleanup error for required property with undefined

jason-ha committed:
test(client): additional Deserialize tests

jason-ha committed:
fix(client): Deserialized objects with recursion

jason-ha committed:
test(client): correct test value for exactOptionalPropertyTypes

jason-ha committed:
fix(client): Deserialized objects with recursion preserved exactly
when all properties are supported.

jason-ha committed:
style(client): apply biome formatting

jason-ha committed:
test(client): use object with getter/setter types
for deserialization tests in place of class instances

jason-ha committed:
test(client): seralize/deserialize test completeness

jason-ha committed:
fix(client): shift general class rejection to serialize side

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
fix(client): preserve property modifiers
by using homomorphic mapping with `as`

jason-ha committed:
fix(client): expect NonNullJsonObjectWith for `object`

jason-ha committed:
docs(client): comment updates

jason-ha committed:
fix(client): corrected array element treatment
and misc formatting improvements

jason-ha committed:
fix(client): deserialized optional markings

jason-ha committed:
test(client): fix sparse array test expectation

jason-ha committed:
feat(client): ignore inaccessible members option

jason-ha committed:
test(client): add number edge cases
and detach expect type from input type T to avoid influencing inference of T

jason-ha committed:
test(client): misc clean up

jason-ha committed:
test(client): test specific function support

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
feat(client): expose JSON support filters as beta
reorganize for documentation and add/correct some docs

jason-ha committed:
improvement(client): switch to namespace
for explicit exposed internal type management

jason-ha committed:
build(client): add nyc to core-interfaces clean script

jason-ha committed:
improvement(client): JsonDeserialized API symmetry with JsonSerializable

jason-ha committed:
test(client): consistent Serializable tests

jason-ha committed:
test(client): doc corrections

jason-ha committed:
test(client): remove commented out, unused code

DLehenbauer commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,549 @@
+ * @system
+ */
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace InternalUtilityTypes {
```
Nit: Consider renaming to 'InternalTypes' for consistency with tree DDS?

```ts
import * as InternalTypes from "./internalTypes.js";
export {
	/**
	 * Contains types used by the API, but which serve mechanical purposes and do not represent semantic concepts.
	 * They are used internally to implement API aspects, but are not intended for use by external consumers.
	 */
	InternalTypes,
};
```

jason-ha replied:
I am expecting that we are going to end up with multiple namespaces in some packages for multiple internal things and the only thing that needs to be consistent (my opinion) is `Internal` prefix on the namespace. None of these should be imported from outside of the package (darn api-extractor); so, I think name that makes sense for internal description is good AND we should be able to arbitrarily change these names without it being considered a breaking change for customers.

BTW tree is also investigating use of explicit namespaces similar to what I used here. I tried the tree pattern show above, but api-extractor is unhappy about `{@link }` references when you rename like that.

DLehenbauer committed:
Fix typo jsonDeserialized.ts

jason-ha committed:
docs(client): ts-doc tweaks

DLehenbauer commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,549 @@
+	 * @beta
+	 * @system
+	 */
+	export type NonSymbolWithOptionalPropertyOf<
```
API: Suggest 'OptionalNonSymbolKeysOf<...>'

DLehenbauer commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,549 @@
+	 * @beta
+	 * @system
+	 */
+	export type NonSymbolWithRequiredPropertyOf<
```
Suggest: 'RequiredNonSymbolKeysOf<...>'

DLehenbauer commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,549 @@
+	 * @beta
+	 * @system
+	 */
+	export type IfAtLeastSometimesDeserializable<T, TException, TTrue, TFalse> =
```
Boy, naming these things is hard.  I think what this does is perform a "shallow" type test for serializability?  (e.g., an object or array are sometimes serializable depending on the types of their contents?)

jason-ha replied:
Definitely for complicated tests - which is most because otherwise we don't need a helper.
I will change this up a bit. I am realizing it is too easy to use incorrectly given its implementation. And I will update comments and perhaps the name.

jason-ha committed:
improvement(client): internal types improvements
renames and a rework for safe use

jason-ha committed:
build(client): use `type` export

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha commented:
@CraigMacomber and/or @anthony-murphy, would you be able to find time to review these changes? (Per Daniel's suggestion as he has run out of time before vacation.)

jason-ha committed:
docs(client): breakout JsonSerializable/Deserialized options

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
merge: 'main' fixup for `@beta` use

CraigMacomber commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,565 @@
+
+/**
+ * Collection of utility types that are not intended to be used/imported
+ * directly outside of this package.
```
Might be worth including in this comment they may not follow semver, since I assume thats the intention (both to communicate it to devs working on/using/reviewing these types but also to customers)

jason-ha replied:
I think the `@system` tag is now sufficiently documented to convey that. LMK if you think otherwise.

CraigMacomber commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,565 @@
+						: /* not an object => */ never
+					: /* function => */ never;
+
+	/**
+	 * Sentinel type for use when marking points of recursion (in a recursive type).
+	 * Type is expected to be unique, though no lengths is taken to ensure that.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export interface RecursionMarker {
+		"recursion here": "recursion here";
+	}
```
It's possible for this to collide with regular data. Picking something faster to compare and collision resistant seems like it would be a good idea. unique symbols seem ideal for that.

jason-ha replied:
I was able to switch to symbols.

CraigMacomber commented on packages/common/core-interfaces/src/jsonDeserialized.ts:
```diff
@@ -0,0 +1,71 @@
+ * @beta
+ */
+export interface JsonDeserializedOptions {
+	/**
+	 * A type that is managed by custom deserialization logic (beyond JSON.parse).
+	 *
+	 * The default value is `never`.
+	 */
+	Replaced: unknown;
```
This is a bit confusing. This seems like a runtime set of options (based on it not being a generic type to take in the parameter), but never is a compile time only thing, and this value isn't optional so how does it have a default anyway?

jason-ha replied:
Breaking type out like this is the guidance for our docs. It is only to be used with JsonDeserialized type param and that is where the default exists. Looked more reasonable on the doc site.

CraigMacomber commented on packages/common/core-interfaces/src/jsonDeserialized.ts:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @beta
+ */
+export type JsonDeserialized<
```
Given this is beta and not internal, I assume there is some public place we intend to use this API? Jsonable is currently `@legacy @alpha`, so this is more public than needed to replace that.

MatrixItem and ISharedSummaryBlock seem to be the only exported APIs using Jsonable. Is this intended for them or maybe some future use-case?

jason-ha replied:
`Jsonable` is `@legacy` so this needs to be `@public` to be able to replace it. (Actually, it is `JsonSerializable` that will replace `Jsonable`. There is no deserialized counterpart in the existing stuff.)
Already noted in the description. :)

CraigMacomber commented on packages/common/core-interfaces/src/jsonDeserialized.ts:
```diff
@@ -0,0 +1,71 @@
+ * When used as a filter to inferred generic `T`, a compile-time error can be
+ * produced trying to assign `JsonDeserialized<T>` to `T`.
+ *
+ * Deserialized JSON never contains `undefined`, `symbol`, or function values.
+ * Object properties with values of those types are absent. So properties
+ * become optional (when there are other supported types in union) or are
+ * removed (when nothing else in union is supported).
+ * In an array, such values are replaced with `null`.
```
When using tools like git which are line based, personally I prefer more semantic line breaks over arbitrary ones mainly
Starting sentences on new lines. Also, I don't think starting a sentence with a coordinating conjunction like "so" like that is correct. either than period should become a comma, or we can use a colon here (which I think makes the most sense):

```suggestion
 * Deserialized JSON never contains `undefined`, `symbol`, or function values.
 * Object properties which had those values before encoding are omitted by JSON.stringify:
 * Therefor in this type properties become optional (when there are other supported types in union)
 * or are removed (when nothing else in union is supported).
 * In an array, such values are replaced with `null`.
```

CraigMacomber commented on packages/common/core-interfaces/src/jsonDeserialized.ts:
```diff
@@ -0,0 +1,71 @@
+ * removed (when nothing else in union is supported).
+ * In an array, such values are replaced with `null`.
+ *
+ * `bigint` valued properties are simply removed as serialization attempts
+ * will throw.
```
There are a lot of things this doc comment does not cover, despite it seeming to make an attempt to be complete, so I'll mention them here:

- Its perfectly possible to losslessly encode bigints into and from JSON. JSON allows arbitrary numeric precision. This doc comment makes no mention of JSON.strinify, so applying its limitation here seems arbitrary. There even is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/rawJSON which makes it possible to to losslesslyencode bigints using JSON.stringify ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON#using_json_numbers )
- No mention of what subset of properties are considered at runtime. Typically, this is enumerable own properties. I thing the section on setters/getters is simply wrong. I'm not sure what you mean by "backed by a function": setters and getters always use functions. Assuming we are talking about JSON.strinify and similar functions what actually matters is if the property is an enumerable own property. Setter don't matter and getters are not special really as they are included when enumerable own. Since TypeScript does not capture ownness or enumerableness of properties in its type system, and most serializers including our stringify and toJson rely on them think this needs to be called out explicitly.
- No mention of the JSON limits on strings or numbers. Strings with unmatched utf16 surrogate pairs will not round trip, nor will NaN or infinities.
- You mention recursive types, but not recursive values

Generally the existing doc comment on Jsonable seems to cover a lot of details which are missing here (ex: the special number cases, assumptions about enumerability of properties, issues with ArrayLike, recursive object issues. If this is intending to replace that API, it seems like its docs should be at least as complete.

CraigMacomber replied:
Maybe this should link JsonSerializable's docs to cover all these cases, and document (at least in a private remarks) why we need both and how they differ?

jason-ha replied:
I've enhanced what it says about what it does without calling out an explicit comparison. And have added limitation notes. Please have another look.

CraigMacomber commented on packages/common/core-interfaces/src/exposedUtilityTypes.ts:
```diff
@@ -0,0 +1,565 @@
+				: /* non-object => T as is */ T;
+
+	/**
+	 * Test for non-public properties (class instance type)
```
I played with this a bit. Its surprisingly robust. Neat utility.

```suggestion
	 * Test for non-public properties (which can only exist on class instance types).
	 * Returns true if `T` deeply may contain (via non function members) a private or protected field.
	 * @remarks
	 * This does not distribute over unions: `ReplaceRecursionWith<A> | ReplaceRecursionWith<B>` will produce Boolean when `ReplaceRecursionWith<A | B>` produces true;
```

I did however find at least one edge case it doesn't handle: classes which extend Function return false even if they have private members. I didn't try out recursive cases.

Might be worth calling out that this has no impact on properties which are present on objects, but not included in the type. That may be obvious, but with the ways this is used, that limitation is likely to impact people. If I implement an interface with a class, or up cast a more specific type to a more general interface before it goes into the JsonSeralizable logic, it all fails to detect the omitted members despite them still impacting runtime. This limitation seems more likely to cause issues most of the things it does detect, which makes me wonder if this whole mechanism is really worth all the complexity as there is no way to fix that issue.

jason-ha replied:
I've added notes. Hopefully it will help people working quick and not being careful about classes. Use of interfaces and hiding classes takes extra effort and my experience says that is not where people start. I became particularly concerned with this issue looking at Bohemia codebase and took a stab at it.

I think I have covered the additional cases and made various improvements as I went.

CraigMacomber commented on packages/common/core-interfaces/src/jsonDeserialized.ts:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @beta
+ */
+export type JsonDeserialized<
```
I see reasonable (but in my opinion not particularly well motivated given the complexity and limitations) use-cases for JsonSserialized, but JsonDeserialized doesn't have a usecase I can find other than casting unvalidated deserialized data to a type it might not actually be compatible with. While I'm ok with customers doing such casts when they think its safe, them doing so does not require this type since they can just cast to their actual type.

Having us use this inside a library seems like its encapsulating a possibly incorrect cast which there is no possible way to validate into our framework, which seems like a bad idea, especially if we expose it publicly. We might have some internal use-cases for this, but I don't think we should use it in a public API, so maybe this specific type should be marked internal instead of beta?

CraigMacomber replied:
The way this API is factored doesn't align well with an API to do to optional validation of data (there is no runtime data format / schema). If you had an API which could do optional validation, you would never use this APIs with it. Tree would never want to use it since we have both internal and user facing schema systems supporting optional data validation and I personally wouldn't make a different design decision elsewhere to use this.

Also is there a need for this to go straight to beta? Could start at alpha.

Regardless: this is entirely a question of API design, I think FF-API really should make the call on if this should exist or not. While I'm on the side of not, it isn't my place to make that subjective API design decision.

The serialization side of this is much less problematic and I have no conceptual opposition to it existing (other than maintenance cost, API surface size and having to stabilize it if we make it non-internal), though if you have code that can safely serialize, it generally comes with an API to serialize so I don't see a use case for it in robust code but its existence as something people will be tempted to use doesn't seem as actively harmful like I consider  JsonDeserialized to be.

jason-ha replied:
`JsonDeserialized` does not make a statement about validity of the type. The utility it provides is to explain what one would get back if they serialized and then deserialized a certain type.
While I didn't add the additional use here, there are additional internal uses where we don't care about validation. The newly motivated case is for multi-process testing where messages are serialized and sent across. During development of that support I noticed that types were mis-authored. Wanted to use this type, but it wasn't available.
As with all tools there are appropriate and inappropriate uses. I believe we have appropriate uses and would like to have this available.

markfields replied:
Do we want to avoid defining types for use with serialization which don't roundtrip through JSON?  i.e. if I need to use `JsonDeserialized`, maybe I'd rather just go update `T` such that `T` is equivalent to `JsonDeserialized<T>`.

Or is this all about the fact that if `T` has a property whose type includes `undefined` (but it's not optional), then on the other side it'll be optional...?  Still, if that matters, then maybe use `null`?

Not a blocking comment saying we shouldn't expose `JsonDeserialized`, just musing about whether it's a stop-gap or whether in the ideal world it would also continue to have a place.

CraigMacomber commented on packages/common/core-interfaces/src/jsonSerializable.ts:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @beta
+ */
+export type JsonSerializable<
```
Given we don't have actual users for this yet, is there a reason for this to skip the alpha stage straight to beta? Seems like this could either start as internal or alpha, and get promoted to beta when the libraries that depend on it need to be marked beta?

I'm generally in favor of not stabilizing things any more than we have a reason to maximize our flexibility and API readability (unused things in API docs make finding things you need harder).

jason-ha replied:
I believe @vladsud has a use and I want it for experimental DIS that is very definitely getting promoted. And I started this a while back. So when I made `@alpha` available I didn't try to demote it. I certainly can.

CraigMacomber commented on packages/common/core-interfaces/src/jsonSerializationErrors.ts:
```diff
@@ -0,0 +1,34 @@
+ * result can be more informative than just the type name.
+ *
+ * @beta
+ * @system
```
do we document `@system` user facing anywhere yet?

jason-ha replied:
We need to add support to the website. @Josmithr, want to give it a go? :) I did add notes to our wiki: https://github.com/microsoft/FluidFramework/wiki/TSDoc-Guidelines#system

jason-ha replied:
Website support added with #22050

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
test(client): remove secondary inference interference
for sparse array test cases T was being promoted to `(number | null)[]` per validation data

jason-ha committed:
test(client): improved failure debugging

jason-ha committed:
test(client): functions with properties and IFluidHandle

jason-ha committed:
improvement(client): internal recursion limit is now unbounded
by making iterator process a string pattern instead of fixed set of number literals

jason-ha committed:
docs(client): remove unused `@beta` tags
from types within namespace

jason-ha committed:
feat(client): distringuish exact from extension allowances

jason-ha committed:
test(client): cleanup stringify undefined result handling and use common bigint replacer/reviver

jason-ha committed:
test(client): function object with recursion

jason-ha committed:
fix(client): recursion nested next to unsupported
Desrialized case yet to be fixed

jason-ha committed:
improvement(client) reduce Deserialization unrolling to 4 times
4 is sufficient for instances to get the idea whereas 10 leads to trouble interpretting the results.

jason-ha committed:
fix(client): recursion nested next to unsupported for Desrialized

jason-ha committed:
fix(client): prevent symbol property from detection as enum-like

jason-ha committed:
test(client): common IFluidHandle testing
Implementaton is failing as IFluidHandle has non-public type - tests have error suppression with TODOs

jason-ha committed:
docs(client): refresh readme with beta use note

jason-ha committed:
docs(client): replace HTML code with logical meaning
linting complains about needing to escape `|` but `\` does not appear to work. "or" is close enough.

jason-ha committed:
improvement(client): update RecursionMarker to use unique symbol

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
improvement(client): support recursion processing through function objects

jason-ha committed:
fix(client): IFluidHandle serialization support
Desrialization with handle and recursion still unrolls

jason-ha committed:
test(client): cover exactOptionalPropertyTypes on and off

jason-ha committed:
merge: 'main' into JsonSerializable-WIP

jason-ha committed:
merge: 'main' into JsonSerializable-WIP

jason-ha committed:
docs(client): small typo/spelling corrections

jason-ha committed:
improvement(client): update Replace*RecursionWith
to use exact matches leveraging new IfExactTypeInTuple
Additionally update some comments and add regions for JsonSerializable and JsonDeserialized exclusive portions

jason-ha committed:
improvement(client): update JsonSerializable support
to use exact matches leveraging new IfExactTypeInTuple

And remove extra allowance for `undefined` as only type for an optional property, when using `exactOptionalPropertyTypes`.

jason-ha committed:
fix(client): support IFluidHandles + recursion
Deserialized recursion support was altering IFluidHandle class type resulting in a mismatch that type overall needed filtering.
Fixed by preserving allowed types while managing recursion.

jason-ha committed:
fix(client): function & object deserialization correctness

Serialization via JSON.stringify() varies for functions intersected with
objects by the runtime type associated with the value. Accomodate this
variance by allowing properties of the object to be possible
deserialized results.

Generally improve public comments as well.

jason-ha committed:
test(client): add bigint array coverage

jason-ha committed:
fix(client): `bigint` special casing
`bigint` is different from other unsupported types as those are mostly silently dropped or made null. For `bigint` there are serialization errors throw. That means that the results are a little different.

Add more deserialization documentation.

jason-ha committed:
docs(client): HasNonPublicProperties
clarity and limitations

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
improvement(client): tighten up union handling
be more explicit about union processing
Get rid of `true` else `false` utilities to avoid potential `boolean` in unexpected context.

jason-ha committed:
docs(client): correct typo

CraigMacomber commented on packages/common/core-interfaces/package.json:
```diff
@@ -87,11 +100,37 @@
 		"typetests:gen": "flub generate typetests --dir . -v",
 		"typetests:prepare": "flub typetests --dir . --reset --previous --normalize"
 	},
+	"c8": {
```
There are a lot of changes in this PR that don't seem related or explained in the PR description. Perhaps you could front load those where possible (ex: add support for beta APIs as a change, and add support for testing in this package as another) so this change could focus on the types. I don't think I'm a great reviewer for all this package setup, script, api extractor and linter changes (at least I can't confidently say they are right, and have a reason to be made as part of this change)

jason-ha replied:
All are PR related.
- The API extractor changes were for adding `@beta` support, which should only be added when there is a `@beta` API. Changes are reduced to `@legacy` and support changes are now removed.
- The test changes are for adding test support. Previously there were no tests here. Can't add test support without having tests. Now noted in PR description.

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
docs(client): update comments

jason-ha committed:
build(client): retarget APIs to legacy

jason-ha committed:
revert(client): remove core-interface `@beta` support

jason-ha committed:
test(client): add CommonJS test as there is difference
with treatment of imported const enums.

Add specialized test logic to account for the difference.

jason-ha committed:
build(client): move Json* APIs internal
to get implementation in code base

All changes are type only changes and thus do not impact production (assuming type files are excluded from packages).

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
test(client): correct comment typo

jason-ha committed:
Merge branch 'main' into JsonSerializable

jason-ha committed:
refactor(client): Json serializable utility types
     in core-interfaces relocated from presence and promoted to beta (though kept hidden under `internal/exposedUtilityTypes` and undocumented)

jason-ha committed:
improvement(client-core-interfaces): internal cleanup
- Add usage docs for importing /internal versus /internal/exposedUtilityType
- Fully specify /internal re-exposure and add JsonTypeWith
- Trim to internal usages to /internal

jason-ha committed:
improvement(client-core-interfaces): self reference support

jason-ha committed:
build(client-core-interfaces): add package.json stub exclusion

jason-ha committed:
Merge branch 'main' into JsonSerializable

jason-ha committed:
docs(client-core-interfaces): correct typo

jason-ha committed:
test(client-core-interfaces): readonly array

jason-ha committed:
test(client-core-interfaces): Map and Set

jason-ha committed:
test(client-core-interfaces): cleanup stale test code

pragya91 commented:
@jason-ha , for _"JsonSerializable should eventually replace @fluidframework/datastore-definitions's Jsonable"_ , we should create a backlog item to track it, with details on which version it would be viable to do so.

markfields commented:
It would be a cool follow-up to have JsonString<T> type that is type branded string that implies that if you parse it you'll get JsonDeserialized<T> or something (not sure exactly how to leverage your types here).  And then we write strongly-typed wrappers for `JSON.stringify` and `JSON.parse` to yield/use `JsonString` so we can have strong typing even after serialization.

jason-ha committed:
test(client-core-interfaces): required lint suppression

jason-ha commented:
> It would be a cool follow-up to have JsonString type that is type branded string that implies that if you parse it you'll get JsonDeserialized or something (not sure exactly how to leverage your types here). And then we write strongly-typed wrappers for `JSON.stringify` and `JSON.parse` to yield/use `JsonString` so we can have strong typing even after serialization.
> 
> I've prototyped this a few times, would be very useful in ContainerRuntime layer where we pass around stringified stuff all the time.

Maybe I can FHL it. Something like `JsonString<T>` should be enough and parser/encoder can qualify what is supported for handles and such in their declarations.
`function encodeWithHandles<T>(v: JsonSerializable<T, { AllowExtensionOf: [IFluidHandle] }>): JsonString<T>`
`function decodeWithHandles<T>(v: JsonString<T>): JsonDeserialized<T, { AllowExtensionOf: [IFluidHandle] }>`
(`JsonDeserialized<T, { AllowExtensionOf: [IFluidHandle] }>` will just be `T` for perfectly round-trippable T.)

jason-ha commented:
> @jason-ha , for _"JsonSerializable should eventually replace @fluidframework/datastore-definitions's Jsonable"_ , we should create a backlog item to track it, with details on which version it would be viable to do so.

Agreed. If the internal uses continue to look good, we could make a change for 3.0. (The Pages codebase may not be able to transition before - there was some cleanup needed when I made fixes to Jsonable a while back.)

jason-ha committed:
test(client-core-interfaces): minor cleanup
- remove extraneous test types
- rename test value for accuracy
- also comment corrections

jason-ha committed:
improvement(client-core-interfaces): AllowExactly: `unknown` support
Separate internal recursion detection use of AllowExactly filter control from user given so that user may specify `unknown`.

A better approach would be to turn AllowExactly into a tuple. Experimentation as needed as in the past, tuple use appeared to cause other issues.

jason-ha committed:
improvement(client-core-interfaces): branded primitive support
Support primitives intersected with classes to be serialized (`JsonSerializable`). (`JsonDeserialized` already had support.)

jason-ha committed:
improvement(client-core-interfaces): AllowExactly as tuple
Update `AllowExactly` option to accept tuple of types that is more flexible than union as it allows `unknown` to be mixed with other types.

jason-ha committed:
improvement(client-core-interfaces): remove InternalAllowExactly
Now that AllowExactly is a tuple ReplacementMaker can be added to it directly instead of requiring a separate control property even when AllowExactly may contain `unknown`.

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
fix(client-core-interfaces): index signature support
especially for those with `unknown` value types.

jason-ha committed:
improvement(client-core-interfaces): better displayed "error" results for users
Instead of some results displaying `FormDegenerate...` use direct formation to have `JsonTypeWith` show up explicitly.

jason-ha committed:
merge: 'main' into JsonSerializable

jason-ha committed:
test(client-core-interfaces): object nested array coverage

jason-ha commented:
> > It would be a cool follow-up to have JsonString type that is type branded string that implies that if you parse it you'll get JsonDeserialized or something (not sure exactly how to leverage your types here). And then we write strongly-typed wrappers for `JSON.stringify` and `JSON.parse` to yield/use `JsonString` so we can have strong typing even after serialization.
> > I've prototyped this a few times, would be very useful in ContainerRuntime layer where we pass around stringified stuff all the time.
> 
> Maybe I can FHL it. Something like `JsonString<T>` should be enough and parser/encoder can qualify what is supported for handles and such in their declarations. `function encodeWithHandles<T>(v: JsonSerializable<T, { AllowExtensionOf: IFluidHandle }>): JsonString<T>` `function decodeWithHandles<T>(v: JsonString<T>): JsonDeserialized<T, { AllowExtensionOf: IFluidHandle }>` (`JsonDeserialized<T, { AllowExtensionOf: [IFluidHandle] }>` will just be `T` for perfectly round-trippable T.)

I worked on this during FHL which found some interesting use cases (branded strings and explicit `unknown`), that resulted in all of the additional changes this last week. That FHL work is not complete, but I don't think there will be needs for additional changes to implementation.
Going to strive to stop making any more changes outside of critical code review feedback.

markfields commented on packages/common/core-interfaces/src/test/jsonDeserialized.spec.ts:
```diff
@@ -415,13 +481,80 @@ describe("JsonDeserialized", () => {
 			});
 
+			it("object with array of supported types (numbers) are preserved", () => {
+				const resultRead = passThru(objectWithArrayOfNumbers);
```
I would have generated the type name based on the variable name, but the fact that you typed out both gives me confidence for how thoroughly you've thought through these test cases!

jason-ha replied:
There are a lot of cases where the name is exactly what is intended to be covered. Then there are some like this one that is meant to handle a set of things but only tests a specific case. We only test array of numbers, but numbers is proxy for other supported things like booleans, strings, and supported objects.

jason-ha commented:
All of the tests:

- JsonDeserialized
  - positive compilation tests
      - supported primitive types are preserved
        -✔ `boolean`
        -✔ `number`
        -✔ `string`
        -✔ numeric enum
        -✔ string enum
        -✔ const heterogenous enum
        -✔ computed enum
        -✔ branded `number`
        -✔ branded `string`
     - unions with unsupported primitive types preserve supported types
        -✔ `string | symbol`
        -✔ `bigint | string`
        -✔ `bigint | symbol`
        -✔ `number | bigint | symbol`
     - supported literal types are preserved
        -✔ `true`
        -✔ `false`
        -✔ `0`
        -✔ "string"
        -✔ `null`
        -✔ object with literals
        -✔ array of literals
        -✔ tuple of literals
        -✔ specific numeric enum value
        -✔ specific string enum value
        -✔ specific const heterogenous enum value
        -✔ specific computed enum value
     - arrays
        -✔ array of supported types (numbers) are preserved
        -✔ sparse array is filled in with null
        -✔ array of partially supported (numbers or undefined) is modified with null
        -✔ array of `unknown` becomes array of `JsonTypeWith<never>`
        -✔ array of partially supported (bigint or basic object) becomes basic object only
        -✔ array of partially supported (symbols or basic object) is modified with null
        -✔ array of unsupported (bigint) becomes never[]
        -✔ array of unsupported (symbols) becomes null[]
        -✔ array of unsupported (functions) becomes null[]
        -✔ array of functions with properties becomes ({...}|null)[]
        -✔ array of objects and functions becomes ({...}|null)[]
        -✔ array of `bigint | symbol` becomes null[]
        -✔ array of `number | bigint | symbol` becomes (number|null)[]
        -✔ readonly array of supported types (numbers) are preserved
     - fully supported object types are preserved
        - ✔ empty object
        - ✔ object with `boolean`
        - ✔ object with `number`
        - ✔ object with `string`
        - ✔ object with number key
        - ✔ object with array of supported types (numbers) are preserved
        - ✔ object with sparse array is filled in with null
        - ✔ object with branded `number`
        - ✔ object with branded `string`
        - ✔ `string` indexed record of `number`s
        - ✔ `string`|`number` indexed record of `string`s
        - ✔ `string` indexed record of `number`|`string`s with known properties
        - ✔ `string`|`number` indexed record of `strings` with known `number` property (unassignable)
        - ✔ `Partial<>` `string` indexed record of `numbers`
        - ✔ templated record of `numbers`
        - ✔ `Partial<>` templated record of `numbers`
        - ✔ object with possible type recursion through union
        - ✔ object with optional type recursion
        - ✔ object with deep type recursion
        - ✔ object with alternating type recursion
        - ✔ simple json (`JsonTypeWith<never>`)
        - ✔ non-const enum
        - ✔ object with `readonly`
        - ✔ object with getter implemented via value
        - ✔ object with setter implemented via value
        - ✔ object with matched getter and setter implemented via value
        - ✔ object with mismatched getter and setter implemented via value
        - class instance
           -✔ with public data (propagated)
        - object with optional property (remains optional)
           -✔ without property
           -✔ with undefined value (property is removed in value)
           -✔ with defined value
     - partially supported object types are modified
        -✔ `object` (plain object) becomes non-null Json object
      - fully unsupported properties are removed
          -✔ object with exactly `bigint`
          -✔ object with exactly `symbol`
          -✔ object with exactly function
          -✔ object with exactly `Function | symbol`
          -✔ object with inherited recursion extended with unsupported properties
          -✔ object with required exact `undefined`
          -✔ object with optional exact `undefined`
          -✔ object with exactly `never`
          -✔ `string` indexed record of `undefined`
          -✔ `string` indexed record of `undefined` and known `number` property (unassignable)
       - partially unsupported properties become optional for those supported
          - ✔ object with exactly `string | symbol`
          - ✔ object with exactly `bigint | string`
          - ✔ object with exactly `bigint | symbol`
          - ✔ object with exactly `number | bigint | symbol`
          - ✔ object with symbol key
          - ✔ object with recursion and `symbol` unrolls 4 times and then has generic Json
          - ✔ object with exactly function with properties
          - ✔ object with exactly object and function
          - ✔ object with function object with recursion
          - ✔ object with object and function with recursion
          - ✔ object with required `unknown` in recursion when `unknown` is allowed unrolls 4 times with optional `unknown`
          - object with `undefined`
            -✔ with undefined value
            -✔ with defined value
       - partially supported array properties are modified like top-level arrays
          -✔ object with array of partially supported (numbers or undefined) is modified with null
          -✔ object with array of `unknown` becomes array of `JsonTypeWith<never>`
          -✔ object with array of partially supported (bigint or basic object) becomes basic object only
          -✔ object with array of partially supported (symbols or basic object) is modified with null
          -✔ object with array of unsupported (bigint) becomes never[]
          -✔ object with array of unsupported (symbols) becomes null[]
          -✔ object with array of unsupported (functions) becomes null[]
          -✔ object with array of functions with properties becomes ({...}|null)[]
          -✔ object with array of objects and functions becomes ({...}|null)[]
          -✔ object with array of `bigint | symbol` becomes null[]
          -✔ object with array of `number | bigint | symbol` becomes (number|null)[]
          -✔ object with readonly array of supported types (numbers) are preserved
       - function & object intersections preserve object portion
          -✔ function with properties
          -✔ object and function
          -✔ function with class instance with private data
          -✔ function with class instance with public data
          -✔ class instance with private data and is function
          -✔ class instance with public data and is function
          -✔ function object with recursion
          -✔ object and function with recursion
       - class instance methods and non-public properties are removed
          - ✔ with public method (removes method)
          - ✔ with private method (removes method)
          - ✔ with private getter (removes getter)
          - ✔ with private setter (removes setter)
          - ✔ with private data (hides private data that propagates)
          - ✔ object with recursion and handle unrolls 4 times listing public properties and then has generic Json
          - for common class instance of
            -✔ Map
            -✔ ReadonlyMap
            -✔ Set
            -✔ ReadonlySet
       - branded non-primitive types lose branding
          -✔ branded `object` becomes just empty
          -✔ branded object with `string`
     - unsupported object types
       - known defect expectations
         - ✔ array of numbers with holes
         - getters and setters preserved but do not propagate
            -✔ object with `readonly` implemented via getter
            -✔ object with getter
            -✔ object with setter
            -✔ object with matched getter and setter
            -✔ object with mismatched getter and setter
   - negative compilation tests
     - assumptions
        -✔ const enums are never readable
     - unsupported types
        -✔ `undefined` becomes `never`
        -✔ `unknown` becomes `JsonTypeWith<never>`
        -✔ `string` indexed record of `unknown` replaced with `JsonTypeWith<never>` (and becomes optional per current TS behavior)
        -✔ templated record of `unknown` replaced with `JsonTypeWith<never>` (and becomes optional per current TS behavior)
        -✔ `string` indexed record of `unknown` and known properties has `unknown` replaced with `JsonTypeWith<never>` (and becomes optional per current TS behavior)
        -✔ `string` indexed record of `unknown` and optional known properties has `unknown` replaced with `JsonTypeWith<never>` (and becomes optional per current TS behavior)
        -✔ `string` indexed record of `unknown` and required known `unknown` has all `unknown` replaced with `JsonTypeWith<never>` (and known becomes explicitly optional)
        -✔ `string` indexed record of `unknown` and optional known `unknown` has all `unknown` replaced with `JsonTypeWith<never>` (and becomes optional per current TS behavior)
        -✔ `Partial<>` `string` indexed record of `unknown` replaced with `JsonTypeWith<never>`
        -✔ `Partial<>` `string` indexed record of `unknown` and known properties has `unknown` replaced with `JsonTypeWith<never>`
        -✔ `symbol` becomes `never`
        -✔ `unique symbol` becomes `never`
        -✔ `bigint` becomes `never`
        -✔ function becomes `never`
        -✔ `void` becomes `never`
   - special cases
     - ✔ explicit `any` generic limits result type
     - using alternately allowed types
       - are preserved
          -✔ `bigint`
          -✔ object with `bigint`
          -✔ object with optional `bigint`
          -✔ array of `bigint`s
          -✔ array of `bigint` or basic object
          -✔ object with specific function
          -✔ `IFluidHandle`
          -✔ object with `IFluidHandle`
          -✔ object with `IFluidHandle` and recursion
          -✔ `unknown`
          -✔ object with optional `unknown`
          -✔ object with optional `unknown` and recursion
          -✔ `string` indexed record of `unknown`
          -✔ templated record of `unknown`
          -✔ `string` indexed record of `unknown` and known properties
          -✔ `string` indexed record of `unknown` and optional known properties
          -✔ `string` indexed record of `unknown` and optional known `unknown`
          -✔ `Partial<>` `string` indexed record of `unknown`
          -✔ `Partial<>` `string` indexed record of `unknown` and known properties
          -✔ array of `unknown`
          -✔ object with array of `unknown`
       - still modifies required `unknown` to become optional
          -✔ object with required `unknown`
          -✔ object with required `unknown` adjacent to recursion
          -✔ mixed record of `unknown`
          -✔ `string` indexed record of `unknown` and required known `unknown`
       - continue rejecting unsupported that are not alternately allowed
          -✔ `unknown` (simple object) becomes `JsonTypeWith<bigint>`
          -✔ `unknown` (with bigint) becomes `JsonTypeWith<bigint>`
          -✔ `symbol` still becomes `never`
          -✔ `object` (plain object) still becomes non-null Json object

- JsonSerializable
   - positive compilation tests
     - supported primitive types
        -✔ `boolean`
        -✔ `number`
        -✔ `string`
        -✔ numeric enum
        -✔ string enum
        -✔ const heterogenous enum
        -✔ computed enum
        -✔ branded `number`
        -✔ branded `string`
     - supported literal types
        -✔ `true`
        -✔ `false`
        -✔ `0`
        -✔ "string"
        -✔ `null`
        -✔ object with literals
        -✔ array of literals
        -✔ tuple of literals
        -✔ specific numeric enum value
        -✔ specific string enum value
        -✔ specific const heterogenous enum value
        -✔ specific computed enum value
     - supported array types
        -✔ array of `number`s
        -✔ readonly array of `number`s
     - supported object types
        - ✔ empty object
        - ✔ object with `never`
        - ✔ object with `boolean`
        - ✔ object with `number`
        - ✔ object with `string`
        - ✔ object with number key
        - ✔ object with array of `number`s
        - ✔ readonly array of `number`s
        - ✔ object with branded `number`
        - ✔ object with branded `string`
        - ✔ `string` indexed record of `number`s
        - ✔ `string`|`number` indexed record of `string`s
        - ✔ templated record of `numbers`
        - ✔ `string` indexed record of `number`|`string`s with known properties
        - ✔ `string`|`number` indexed record of `strings` with known `number` property (unassignable)
        - ✔ object with possible type recursion through union
        - ✔ object with optional type recursion
        - ✔ object with deep type recursion
        - ✔ object with alternating type recursion
        - ✔ simple json (JsonTypeWith<never>)
        - ✔ non-const enums
        - ✔ object with `readonly`
        - ✔ object with getter implemented via value
        - ✔ object with setter implemented via value
        - ✔ object with matched getter and setter implemented via value
        - ✔ object with mismatched getter and setter implemented via value
        - class instance
          - ✔ with public data (just cares about data)
          - with `ignore-inaccessible-members`
            -✔ with private method ignores method
            -✔ with private getter ignores getter
            -✔ with private setter ignores setter
        - object with optional property
          -✔ without property
          -✔ with undefined value
          -✔ with defined value
     - unsupported object types
       - ✔ object with self reference throws on serialization
       - known defect expectations
         - ✔ sparse array of supported types
         - ✔ object with sparse array of supported types
         - getters and setters allowed but do not propagate
            -✔ object with `readonly` implemented via getter
            -✔ object with getter
            -✔ object with setter
            -✔ object with matched getter and setter
            -✔ object with mismatched getter and setter
         - class instance
           - with `ignore-inaccessible-members`
              -✔ with private data ignores private data (that propagates)
   - negative compilation tests
     - assumptions
        -✔ const enums are never readable
     - unsupported types cause compiler error
        - ✔ `undefined`
        - ✔ `unknown`
        - ✔ `symbol`
        - ✔ `unique symbol`
        - ✔ `bigint`
        - ✔ function
        - ✔ function with supported properties
        - ✔ object and function
        - ✔ object with function with supported properties
        - ✔ object with object and function
        - ✔ function with class instance with private data
        - ✔ function with class instance with public data
        - ✔ class instance with private data and is function
        - ✔ class instance with public data and is function
        - ✔ `object` (plain object)
        - ✔ `void`
        - ✔ branded `object`
        - ✔ branded object with `string`
        - unions with unsupported primitive types
          -✔ `string | symbol`
          -✔ `bigint | string`
          -✔ `bigint | symbol`
          -✔ `number | bigint | symbol`
        - array
          -✔ array of `bigint`s
          -✔ array of `symbol`s
          -✔ array of `unknown`
          -✔ array of functions
          -✔ array of functions with properties
          -✔ array of objects and functions
          -✔ array of `number | undefined`s
          -✔ array of `bigint` or basic object
          -✔ array of `symbol` or basic object
          -✔ array of `bigint | symbol`s
          -✔ array of `number | bigint | symbol`s
       - object
          - ✔ object with exactly `bigint`
          - ✔ object with optional `bigint`
          - ✔ object with exactly `symbol`
          - ✔ object with optional `symbol`
          - ✔ object with exactly `function`
          - ✔ object with exactly `Function | symbol`
          - ✔ object with exactly `string | symbol`
          - ✔ object with exactly `bigint | string`
          - ✔ object with exactly `bigint | symbol`
          - ✔ object with exactly `number | bigint | symbol`
          - ✔ object with array of `bigint`s
          - ✔ object with array of `symbol`s
          - ✔ object with array of `unknown`
          - ✔ object with array of functions
          - ✔ object with array of functions with properties
          - ✔ object with array of objects and functions
          - ✔ object with array of `number | undefined`s
          - ✔ object with array of `bigint` or basic object
          - ✔ object with array of `symbol` or basic object
          - ✔ object with array of `bigint | symbol`s
          - ✔ object with symbol key
          - ✔ `string` indexed record of `unknown`
          - ✔ `Partial<>` `string` indexed record of `unknown`
          - ✔ `Partial<>` `string` indexed record of `numbers`
          - ✔ `Partial<>` templated record of `numbers`
          - ✔ object with recursion and `symbol`
          - ✔ function object with recursion
          - ✔ object and function with recursion
          - ✔ nested function object with recursion
          - ✔ nested object and function with recursion
          - ✔ object with inherited recursion extended with unsupported properties
          - object with `undefined`
            -✔ as exact property type
            -✔ in union property
            -✔ as exact property type of `string` indexed record
            -✔ as exact property type of `string` indexed record intersected with known `number` property (unassignable)
            -✔ as optional exact property type > varies by exactOptionalPropertyTypes setting
            -✔ under an optional property
          - object with required `unknown` even though exactly allowed
            -✔ as exact property type
            -✔ as exact property type adjacent to recursion
            -✔ as exact property type in recursion
          - of class instance
            -✔ with private data
            -✔ with private method
            -✔ with private getter
            -✔ with private setter
            -✔ with public method
       - common class instances
          -✔ Map
          -✔ ReadonlyMap
          -✔ Set
          -✔ ReadonlySet
   - special cases
     - ✔ explicit `any` generic still limits allowed types
     - `number` edge cases
       - supported
          -✔ MIN_SAFE_INTEGER
          -✔ MAX_SAFE_INTEGER
          -✔ MIN_VALUE
          -✔ MAX_VALUE
       - resulting in `null`
          -✔ NaN
          -✔ +Infinity
          -✔ -Infinity
     - using alternately allowed types
       - are supported
          -✔ `bigint`
          -✔ object with `bigint`
          -✔ object with optional `bigint`
          -✔ array of `bigint`s
          -✔ array of `bigint` or basic object
          -✔ object with specific alternately allowed function
          -✔ `IFluidHandle`
          -✔ object with `IFluidHandle`
          -✔ object with `IFluidHandle` and recursion
          -✔ `unknown`
          -✔ array of `unknown`
          -✔ object with array of `unknown`
          -✔ object with optional `unknown`
          -✔ `string` indexed record of `unknown`
          -✔ templated record of `unknown`
          -✔ `string` indexed record of `unknown` and known properties
          -✔ `string` indexed record of `unknown` and optional known properties
          -✔ `string` indexed record of `unknown` and optional known `unknown`
          -✔ `Partial<>` `string` indexed record of `unknown`
          -✔ `Partial<>` `string` indexed record of `unknown` and known properties
          -✔ object with optional `unknown` adjacent to recursion
          -✔ object with optional `unknown` in recursion
       - continue rejecting unsupported that are not alternately allowed
          -✔ `unknown` (simple object) expects `JsonTypeWith<bigint>`
          -✔ `unknown` (with bigint) expects `JsonTypeWith<bigint>`
          -✔ `symbol` still becomes `never`
          -✔ `object` (plain object) still becomes non-null Json object
          -✔ object with non-alternately allowed too generic function
          -✔ object with non-alternately allowed too input permissive function
          -✔ object with non-alternately allowed more restrictive output function
          -✔ object with supported or non-supported function union
          -✔ `string` indexed record of `unknown` and required known `unknown` that must be optional
          -✔ mixed record of `unknown`

- JsonSerializable under exactOptionalPropertyTypes=true
   - negative compilation tests
     - unsupported types cause compiler error
       - object
         - object with `undefined`
           - ✔ as optional exact property type

markfields reviewed (approved):
Disclaimer: I did not review the 1000 lines of metaprogramming, but the breadth and depth of test cases gives me a lot of confidence.  And as discussed, the risks of something bad happening in production here are low - if someone thinks they're serializing something but it's not happening right and the type allows it, they should have tests to cover the runtime behavior.
